training/splitting_dataset.py

```diff
@@ -0,0 +1,132 @@
+    :param dataframe: Processed dataframe.
+    :return: Dictionary containing list of books for each year.
+    """
+    with open("../training/book_year_mapping.json", "r") as mapping_file:
+        book_year_mapping = json.loads("".join(mapping_file.readlines()))
+
+    grouped_books_per_year = defaultdict(list)
+    book_names = _get_book_names_from_dataframe(dataframe)
+    for book_name in book_names:
+        if book_year := book_year_mapping.get(book_name):
+            grouped_books_per_year[book_year].append(book_name)
+        else:
+            grouped_books_per_year[0].append(book_name)
+    return grouped_books_per_year
+
+
+def _get_dataframe_with_oldest_books(
+    dataframe: DataFrameType, test_size: float = _DEFAULT_TEST_SIZE
+) -> DataFrameType:
+    """
+    Extracts from dataframe rows with the oldest books. Number of extracted rows is equal to
+    a fraction of initial number of rows defined in test_size.\n
+    :param dataframe: Processed dataframe.
+    :param test_size: Fraction of the dataset to include in test split. It should be a float number between 0.0 and 1.0.
+    :return: Dataframe containing extracted rows.
+    """
+    books_per_year = _get_grouped_book_names(dataframe)
+    filtered_dataframe = pd.DataFrame(columns=dataframe.columns)
+    remaining_rows_num = int((1 - test_size) * dataframe.shape[0])
+    for year in sorted(books_per_year.keys()):
+        for book_name in books_per_year[year]:
+            book_df = dataframe.loc[dataframe["book"] == book_name]
+            included_rows_num = min(remaining_rows_num, book_df.shape[0])
+            filtered_dataframe = pd.concat(
+                [filtered_dataframe, book_df[:included_rows_num]]
+            )
+            remaining_rows_num -= included_rows_num
+            if remaining_rows_num == 0:
+                return filtered_dataframe
+    return filtered_dataframe
+
+
+def _get_chronological_split_results(
+    dataframe: DataFrameType, test_size: float = _DEFAULT_TEST_SIZE
+) -> Tuple[DataFrameType, DataFrameType, DataFrameType, DataFrameType]:
+    """
+    Splits dataframe into training and testing sets chronologically.\n
+    :param dataframe: Processed dataframe.
+    :param test_size: Fraction of the dataset to include in test split. It should be a float number between 0.0 and 1.0.
+    :return: Two dataframes (first one for training and second one for testing) and two columns in the same order.
+    """
+    X_train = _get_dataframe_with_oldest_books(dataframe, test_size)
+    if not X_train.empty:
+        X_test = dataframe.drop(X_train.index)
+        y_train = X_train.pop("level")
+        y_test = X_test.pop("level")
+        return X_train, X_test, y_train, y_test
+    else:
+        print(
+            "_get_chronological_split_results: No books were found in dataframe. Dataframe will be split randomly."
+        )
+        return _get_random_split_results(dataframe)
```

Comment:
IMO to jest typowy error - użytkownik chciał podzielić chronologicnzie, ale zabrakło potrzebnych danych. Rzucałbym tu wyjątkiem i tyle